src/mem_table.rs

```diff
@@ -248,223 +304,142 @@ impl KVTable {
-        let kv = iter.next().await.unwrap().unwrap();
-        assert_eq!(kv.key, b"abc555".as_slice());
-        assert_eq!(kv.value, b"value5".as_slice());
-        assert!(iter.next().await.unwrap().is_none());
+        table.put(RowEntry::new_value(b"abc333", b"value3", 1));
+        table.put(RowEntry::new_value(b"abc111", b"value1", 2));
+        table.put(RowEntry::new_value(b"abc555", b"value5", 3));
+        table.put(RowEntry::new_value(b"abc444", b"value4", 4));
+        table.put(RowEntry::new_value(b"abc222", b"value2", 5));
+
+        let mut iter = table
+            .table()
+            .range(BytesRange::from(Bytes::from_static(b"abc334")..));
+        assert_iterator(
+            &mut iter,
+            vec![
+                RowEntry::new_value(b"abc444", b"value4", 4),
+                RowEntry::new_value(b"abc555", b"value5", 3),
+            ],
+        )
+        .await;
     }
 
     #[tokio::test]
     async fn test_memtable_iter_delete() {
         let mut table = WritableKVTable::new();
-        table.put(
-            Bytes::from_static(b"abc333"),
-            Bytes::from_static(b"value3"),
-            gen_attrs(1),
-        );
-        table.delete(Bytes::from_static(b"abc333"), gen_attrs(2));
-
-        let mut iter = table.table().iter();
-        assert!(iter.next().await.unwrap().is_none());
+        table.put(RowEntry::new_tombstone(b"abc333", 2));
+        table.put(RowEntry::new_value(b"abc333", b"value3", 1));
+        table.put(RowEntry::new_value(b"abc444", b"value4", 4));
+
+        // in merge iterator, it should only return one entry
+        let iter = table.table().iter();
+        let mut merge_iter = MergeIterator::new(VecDeque::from(vec![iter]))
+            .await
+            .unwrap();
+        assert_iterator(
+            &mut merge_iter,
+            vec![
+                RowEntry::new_tombstone(b"abc333", 2),
+                RowEntry::new_value(b"abc444", b"value4", 4),
+            ],
+        )
+        .await;
     }
 
     #[tokio::test]
     async fn test_memtable_track_sz() {
         let mut table = WritableKVTable::new();
 
         assert_eq!(table.table.size(), 0);
-        table.put(
-            Bytes::from_static(b"first"),
-            Bytes::from_static(b"foo"),
-            gen_attrs(1),
-        );
-        assert_eq!(table.table.size(), 16); // first(5) + foo(3) + attrs(8)
-
-        // ensure that multiple deletes keep the table size stable
-        for ts in 2..5 {
-            table.delete(Bytes::from_static(b"first"), gen_attrs(ts));
-            assert_eq!(table.table.size(), 13); // first(5) + attrs(8)
-        }
+        table.put(RowEntry::new_value(b"first", b"foo", 1));
+        assert_eq!(table.table.size(), 16);
+
+        table.put(RowEntry::new_tombstone(b"first", 2));
+        assert_eq!(table.table.size(), 29);
+
+        table.put(RowEntry::new_tombstone(b"first", 2));
+        assert_eq!(table.table.size(), 29);
```

Comment:
🤔 Is there any case that this is actually a legit call? Should we panic here since it might signal a bigger issue?

Reply:
make sense :+1: